Use picolibc instead of newlib-nano

keith-packard opened the pull request:
1. Add picolibc bits to linker scripts. Picolibc uses thread local storage
        for unshared data, and the picolibc startup code relies on carefully
        constructed linker script that allocates one TLS block in RAM for
        applications that don't have explicit TLS support. The picolibc
        startup code also uses different symbols than the tf-m startup code.
    
     2. Support picolibc stdio. There was already picolibc support code
        present in the library for the ARM LLVM toolchain. The
        conditionals which selected it have been changed to use
        __PICOLIBC__ instead of __clang_major__.
    
     3. Add _exit stub. Code using assert or abort end up calling _exit
        through the picolibc signal handling code.
    
     4. Switch to picolibc.specs. This is needed for toolchains which
        don't use picolibc by default, and can also be used without
        trouble in toolchains where picolibc is the default.

keith-packard committed:
[zep noup] Use picolibc instead of newlib-nano

 1. Support picolibc stdio. There was already picolibc support code
    present in the library for the ARM LLVM toolchain. The
    conditionals which selected it have been changed to use
    __PICOLIBC__ instead of __clang_major__.

 2. Add _exit stub. Code using assert or abort end up calling _exit
    through the picolibc signal handling code.

 3. Switch to picolibc.specs. This is needed for toolchains which
    don't use picolibc by default, and can also be used without
    trouble in toolchains where picolibc is the default.

 4. Define CONFIG_PICOLIBC when using picolibc. This is enabled
    by default, but can be disabled on the cmake command line.

 5. Define picolibc_startup to initialize all .data and .bss
    segments using the standard tf-m arrays.

 6. Add picolibc.c to many source lists when CONFIG_PICOLIBC is
    defined. If this file is missign, picolibc_startup will
    not be defined which should cause a failure at link time.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
[zep noup] ld: Reset linker '.' value after .ER_CODE_SRAM section

When processing the .ER_CODE_SRAM section, the location counter is set
to the VMA of that section, so after processing the next section will
use that VMA instead of the next available address in the FLASH
region.

Reset the location counter to the next available FLASH address after
processing this section will ensure that any subsequent FLASH data
is placed in the correct location.

This test fails without this patch:

$ west build -p -b lpcxpresso55s69/lpc55s69/cpu0/ns \
  samples/synchronization -T sample.kernel.synchronization

.../ld: address 0x14003e84 of bin/tfm_s.axf section `.ER_TFM_CODE'
    is not within region `FLASH'
.../ld: address 0x14005c20 of bin/tfm_s.axf section `.TFM_UNPRIV_CODE'
    is not within region `FLASH'
.../ld: bin/tfm_s.axf section `.TFM_PSA_ROT_LINKER_DATA'
    will not fit in region `FLASH'
.../ld: address 0x14003e84 of bin/tfm_s.axf section `.ER_TFM_CODE'
    is not within region `FLASH'
.../ld: address 0x14005c20 of bin/tfm_s.axf section `.TFM_UNPRIV_CODE'
    is not within region `FLASH'
.../ld: ERROR: CMSE stub (.gnu.sgstubs section) too far (0x10008540)
    from destination (0x14003ee4)

Note that the linker is generating addresses within the CODE_RAM
memory region rather than the FLASH region.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
These are (finally) passing Zephyr tests using both SDK 0.17 and SDK 0.18-alpha4 (although each requires some additional changes which are wending their way upstream). I think they're ready for review and help getting them upstream.

tomi-font commented on platform/ext/common/gcc/tfm_common_s.ld.template:
```diff
@@ -192,6 +192,9 @@ SECTIONS
         . = ALIGN(4); /* This alignment is needed to make the section size 4 bytes aligned */
     } > CODE_RAM AT > FLASH
 
+    /* Reset current position for subsequent sections */
+    . = LOADADDR(.ER_CODE_SRAM) + SIZEOF(.ER_CODE_SRAM);
```
How did the linker get confused exactly? Is this really needed?

tomi-font reviewed (commented):
Had a chat with upstream about these (picolibc) changes. Let's keep them only in this fork until they're not needed anymore (TF-M is moving away from C library usage), in the next release or the one following it.
But it would be good to have someone who understands this a bit to review the changes. Don't know who that could be. cc @frkv @d3zd3z @ceolin @dleach02 @Vge0rge

keith-packard commented:
> Had a chat with upstream about these (picolibc) changes. Let's keep them only in this fork until they're not needed anymore (TF-M is moving away from C library usage), in the next release or the one following it. But it would be good to have someone who understands this a bit to review the changes. Don't know who that could be. cc @frkv @d3zd3z @ceolin @dleach02 @Vge0rge

Sounds good. I'll keep running tests as this series moves forward. This is *not* buildable using SDK 0.17.2 because of a last-minute oops in that release, but it should be buildable with the upcoming SDK 0.17.3.

frkv commented:
I recommend @wearyzen and @adeaarm give it a quick look from TF-M's perspective. 

The [Clang/LLVM toolchain](https://lists.trustedfirmware.org/archives/list/tf-m@lists.trustedfirmware.org/thread/RLGP7YUJMON4V2YTODCJJTMZHOBFMW6T/) for TF-M was added with no dependency on libc and/or picolib in the core TF-M code, but you still see references to nano.specs in the [GNUARM toolchain](https://git.trustedfirmware.org/plugins/gitiles/TF-M/trusted-firmware-m.git/+/refs/heads/release/2.2.x/toolchain_GNUARM.cmake#110)

Zephyr, making this choice for GNUARM toolchain (this change), will take any penalty related to support, any security advisories etc... But the choice to go to picolibc is an active choice in the Zephyr community, and in extension it will impact any vendor who makes use of standard c library function in  PSA crypto and PSA crypto driver build as well as in bootloaders...

tomi-font commented:
> With regards to these commits, I think we should consider them to be marked as commits emitted from zephyr integration

Yeah this is something we've agreed on and I think just needs to be implemented (documenting the practice basically). In the meantime I don't think we can require that of anyone yet.

Anton-TF commented:
True, TF-M plans to remove `libc` from all secure binaries (BL1, BL2, tfm_s) for Clang/LLVM and GCC toolchains at least.
As part of this effort, we also intend to clean up and streamline the linker scripts.

Additionally, you might want to consider these recent fixes in the scripts:
https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/40014
https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/40068

keith-packard commented:
Sounds like any libc-specific changes will be transient then, which will be nice.

wearyzen commented:
While trying to build the non-secure MPS4 board with [this](https://github.com/zephyrproject-rtos/zephyr/pull/95105) PR and Zephyr SDK 0.18.0-alpha4 we see below error:
```
west build -p -b mps4/corstone320/fvp/ns samples/hello_world/ -t run
...
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: (.text._start+0x30): undefined reference to `__data_source'
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: (.text._start+0x34): undefined reference to `__data_start'
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: (.text._start+0x38): undefined reference to `__bss_size'
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: (.text._start+0x3c): undefined reference to `__bss_start'
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: (.text._start+0x40): undefined reference to `__tls_base'
/home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/bin/ld: /home/zephyr-sdk-0.18.0-alpha4/gnu/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/14.3.0/../../../../arm-zephyr-eabi/lib/thumb/v8-m.main/nofp/space/libc.a(libc_picolib_machine_arm_set_tls.c.o): in function `_set_tls':
set_tls.c:(.text._set_tls+0xc): undefined reference to `__arm32_tls_tcb_offset'
```

Referring to the changes in this PR, I tried below patch and it seems to fix the issue:

```
git diff
diff --git a/bl1/bl1_2/bl1_dummy_rotpk.prv b/bl1/bl1_2/bl1_dummy_rotpk.prv
index bf011cd4d..b6bf9b600 100644
Binary files a/bl1/bl1_2/bl1_dummy_rotpk.prv and b/bl1/bl1_2/bl1_dummy_rotpk.prv differ
diff --git a/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_1.ld b/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_1.ld
index 44f04021a..7fad80960 100644
--- a/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_1.ld
+++ b/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_1.ld
@@ -160,6 +160,25 @@ SECTIONS
     } > RAM

     bss_size = __bss_end__ - __bss_start__;
+    /* --- add after the .bss section --- */
+    PROVIDE( __bss_start = __bss_start__ );
+    PROVIDE( __bss_end   = __bss_end__ );
+    PROVIDE( __bss_size  = __bss_end - __bss_start );
+
+    PROVIDE( __data_start  = ADDR(.data) );
+    PROVIDE( __data_end    = __data_end__ );
+    PROVIDE( __data_source = LOADADDR(.data) );
+    PROVIDE( __data_size   = __data_end - __data_start );
+
+    /* BL1_1 doesn’t use TLS, but picolibc’s _set_tls needs these present */
+    PROVIDE( __tls_base = 0 );
+    PROVIDE( __tls_align = 8 );
+    PROVIDE( __arm32_tls_tcb_offset = 8 );
+
+    /* --- add after the .heap section (where __HeapBase/Limit are set) --- */
+    PROVIDE( __heap_start = __HeapBase );
+    PROVIDE( __heap_end   = __HeapLimit );
+

 #if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
     .msp_stack (NOLOAD) : ALIGN(32)
diff --git a/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_2.ld b/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_2.ld
index 83aaf144f..9151ff3a0 100644
--- a/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_2.ld
+++ b/platform/ext/target/arm/mps4/common/device/source/gcc/mps4_corstone3xx_bl1_2.ld
@@ -162,6 +162,25 @@ SECTIONS

     bss_size = __bss_end__ - __bss_start__;

+    /* --- add after the .bss section --- */
+    PROVIDE( __bss_start = __bss_start__ );
+    PROVIDE( __bss_end   = __bss_end__ );
+    PROVIDE( __bss_size  = __bss_end - __bss_start );
+
+    PROVIDE( __data_start  = ADDR(.data) );
+    PROVIDE( __data_end    = __data_end__ );
+    PROVIDE( __data_source = LOADADDR(.data) );
+    PROVIDE( __data_size   = __data_end - __data_start );
+
+    /* BL1_1 doesn’t use TLS, but picolibc’s _set_tls needs these present */
+    PROVIDE( __tls_base = 0 );
+    PROVIDE( __tls_align = 8 );
+    PROVIDE( __arm32_tls_tcb_offset = 8 );
+
+    /* --- add after the .heap section (where __HeapBase/Limit are set) --- */
+    PROVIDE( __heap_start = __HeapBase );
+    PROVIDE( __heap_end   = __HeapLimit );
+
 #if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
     .msp_stack (NOLOAD) : ALIGN(32)
     {
(END)
```

I am not a linker script expert but if above change looks good then please let me know if you would like to include it in this PR.
Thanks!

wearyzen commented:
> @wearyzen could you review this?

I don't have expertise on toolchain or linker so hard for me to say if there is an issue with this but I remember from my [previous comment](https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/134#issuecomment-3234808880) this change is needed to use with Zephyr SDK 0.18.0-alpha4.

Vge0rge reviewed (changes requested):
NACKing until the issues with the Nordic devices are solved

keith-packard commented:
> And this allowed me to at least boot the PSA arch tests and run some crypto test cases. I am investigating more at the moment but in the mean time @keith-packard do you have any ideas how the picolibc can affect something like this?

It could be thread local storage (TLS) stuff. The picolibc changes attempt to stuff a static TLS segment (both `.tdata` and `.tbss`) in between the usual data-initialized memory area (`.data`) and zero-initialized memory area (`.bss`). If any of the symbols that define how those segments are arranged in memory get messed up, memory will not be correctly initialized.

I admit that this is not how thread local storage is usually managed; it's a "clever hack" which avoids needing any additional code to perform that initialization, aside from setting the TLS base pointer.

It could also be other symbols used by picolibc's crt0 for memory initialization aren't getting set correctly.

If you have a linker map file from a failing build, please post it and we can look at the various symbol values and see what's wrong.

If it's the TLS stuff, I'm wondering if we can just assert that TF-M doesn't use any C library functions which need thread local storage -- the most obvious use is for `errno`, but there are a few others used by APIs which are required to be re-entrant and which are also required to hold persistent data.

Vge0rge commented:
@keith-packard  I attached here the linker for when building the TF-M regression tests for nRF54L15.

[nrf54l_tfm_s.txt](https://github.com/user-attachments/files/23052302/nrf54l_tfm_s.txt)

The symbols which I confirmed that they are not initialized correctly are:

- fs_cfg_its
- fs_cfg_ps
- nv_increment_status

There are probably more, I just found these until now. But these are normal global static variables which are placed in .data as expected.  So maybe it is indeed crt related and not TLS after all?

keith-packard committed:
[zep noup] platform: Add picolibc init and heap symbols to linker scripts

Picolibc uses different names for the list of constructors and heap
limits. Add values for __bothinit_array_start, __bothinit_array_end,
__heap_start and __heap_end to every linker script. If not using
picolibc, these will not have any effect.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
and pushed another small set of changes that just stops using picolibc's crt0 entirely; we don't need any CPU initialization it might do, we just need memory set up and constructors run.

Vge0rge commented:
Hey @keith-packard, thanks a lot for the detailed investigation!  I tried the new version of this PR and I still could not boot TF-M in Nordic devices. I did some debugging and I can now confirm that the code to copy tables is now executed form picolib as you described. The issue that I found now is that the linker script does not include the TFM_BSS section in the zero table, so I added this patch:
```
diff --git a/platform/ext/common/gcc/tfm_common_s.ld.template b/platform/ext/common/gcc/tfm_common_s.ld.template
index 8fd95bdea..e623e621a 100644
--- a/platform/ext/common/gcc/tfm_common_s.ld.template
+++ b/platform/ext/common/gcc/tfm_common_s.ld.template
@@ -275,6 +275,9 @@ SECTIONS

         LONG (ADDR(.TFM_APP_ROT_LINKER_BSS))
         LONG (SIZEOF(.TFM_APP_ROT_LINKER_BSS) / 4)
+
+        LONG (ADDR(.TFM_BSS))
+        LONG (SIZEOF(.TFM_BSS) / 4)
 #if defined(CONFIG_TFM_PARTITION_META)
         LONG (ADDR(.TFM_SP_META_PTR))
         LONG (SIZEOF(.TFM_SP_META_PTR) / 4)
```

And this allowed me to boot and run the tests, do you think that you can include this patch in your commits?

keith-packard committed:
[zep noup] platform: Include TFM_BSS in zero table for link script templates

While the copy table included the default data segment for these
templates, the zero table didn't include the default bss segment.

Make sure that area gets initialized to zero using the table-based
mechanism instead of relying on the startup code to initialize it
explicitly.

Signed-off-by: Keith Packard <keithp@keithp.com>

Vge0rge reviewed (commented):
Thanks again for the fix @keith-packard . I can now confirm that this patch works for Nordic devices :) 

Feel free to rework the history and let me know when you are done so that I can approve this. 

Small suggestion, it would be nice in my opinion to include tags like the one used in mbedtls fork of Zephyr (https://github.com/zephyrproject-rtos/mbedtls/blob/zephyr/README.md) when you do the rewrite. I am not gonna block the PR for this since this is not an official rule in this repo yet but I am just suggesting it.

Edit: I ended up approving so that I don't NACK you anymore. If you rewrite it will anyway request my review again.

tomi-font reviewed (commented):
Can you please add details to the commit message of at least the first commit to briefly explain why this is not submitted upstream and what is expected from upstream related to that? for tracking purposes

Vge0rge commented:
D

> > Thanks again for the fix @keith-packard . I can now confirm that this patch works for Nordic devices :)
> > Feel free to rework the history and let me know when you are done so that I can approve this.
> 
> Just pushed a rebase -i that cleans up the sequence a bit and should work through bisect, but the result is exactly the same as what you've already tested (at least according to git diff?)
> 
> > Small suggestion, it would be nice in my opinion to include tags like the one used in mbedtls fork of Zephyr (https://github.com/zephyrproject-rtos/mbedtls/blob/zephyr/README.md) when you do the rewrite. I am not gonna block the PR for this since this is not an official rule in this repo yet but I am just suggesting it.
> 
> Sure, happy to tag these with [zep noup] as we're expecting upstream to drop libc entirely, making this whole sequence unnecessary.
> 
> > Edit: I ended up approving so that I don't NACK you anymore. If you rewrite it will anyway request my review again.
> 
> Thanks. Really appreciate your help getting this over the line. I'm looking forward to having SDK 1.0 fully supported by Zephyr 4.3, even if it won't be the "official" toolchain.

Diff looks good to me as well :) Thanks for adding the tags and for the support as well. Glad to see that moving forward!

etienne-lms reviewed (commented):
I've successfully tested on the few supported STM32 platforms.

For info, while a couple of hundred bytes are saved in flash and RAM when using picolib, I also saw that BL2 footprint was increased by more than 2kByte of flash when console traces are enabled. Observed (from build tests) on various platforms, not only the STM32 ones.

keith-packard commented on platform/ext/common/picolibc.c:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2025, Keith Packard <keithp@keithp.com>
```
(C) is meaningless. I'll just delete the ©

keith-packard commented:
> For info, while a couple of hundred bytes are saved in flash and RAM when using picolib, I also saw that BL2 footprint was increased by more than 2kByte of flash when console traces are enabled. Observed (from build tests) on various platforms, not only the STM32 ones.

What are console traces and how are they enabled? There's nothing obvious in the source tree that I could  find. 

If this ends up enabling 'assert' macros, then that can cause a pretty significant jump in flash usage as the default picolibc assert behavior includes `__FILE__`, `__LINE__` and also the asserting expression as a string. Zephyr SDK 1.0 changes that default to remove that information, so asserts no longer cause significant increases in flash usage.

I just reviewed the picolibc code and there's no easy way to disable the verbose assert behavior once enabled. If you want to do this, you'd need to do:

```
/* make sure the library defaults are loaded */
#include <stdlib.h>
/* unset any default for __ASSERT_VERBOSE */
#undef __ASSERT_VERBOSE
/* get the quiet version of assert defined */
#include <assert.h>
```

d3zd3z reviewed (commented):
Overall, I think this is a reasonable, and hopefully temporary workaround. As mentioned, TF-M is trying to move away from libc, so we probably can't upstream this. It is kind of frustrating how dependent this is on the SDK, but I'm not sure that can be avoided.

keith-packard commented:
Rebased atop current zephyr main branch and re-indented the picolibc.c file using four spaces.

tpambor reviewed (approved):
LGTM.

I have tested this PR on STM32H573I-DK and works as expected.

```
[INF] Starting bootloader
[WRN] This device was provisioned with dummy keys. This device is NOT SECURE
[INF] Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
[INF] Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[INF] Boot source: primary slot
[INF] Image index: 1, Swap type: none
[INF] Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
[INF] Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
[INF] Boot source: primary slot
[INF] Image index: 0, Swap type: none
[INF] Bootloader chainload address offset: 0x38000
[INF] Image version: v0.0.0
[INF] Jumping to the first image slot
Booting TF-M v2.2.0+d491c6136
[WRN] This device was provisioned with dummy keys. This device is NOT SECURE
[Sec Thread] Secure image initializing!
[INF][PS] Encryption alg: 0x5500200
[INF][Crypto] Init HW accelerator...
[INF][Crypto] Init HW accelerator... complete.
*** Booting Zephyr OS build v4.3.0-1147-g51131b46e396 ***
TF-M IPC on stm32h573i_dk
The version of the PSA Framework API is 257.
The PSA Crypto service minor version is 1.
Generating 256 bytes of random data:
DD F4 89 1F 0D DF 40 0E D5 7F D1 44 1E 71 48 C9
05 9C 89 05 0F 63 FA B4 FB BA A3 3A 34 8F 54 AF
9A 1D 30 00 BF D5 AB 57 A5 45 B5 8D 9E 14 CC 38
DC 24 04 6D ED DA BA 3B F0 98 D0 A3 65 73 15 E6
EE 1C F0 75 7D D8 6B 9E B8 88 46 15 61 7B 30 E4
8F BB 2A 95 3F 2C EE D9 75 84 2B 0F 7D 14 88 E2
DA 13 10 22 4C FA 1B 73 D9 DE 61 6E 8F 7A 17 91
BC 1E 8D 14 8D C9 5F B7 4C C7 49 05 2C 3B B5 C6
F3 D7 EB 53 BD E2 00 54 E4 8E 4A 4D 84 D9 88 B9
A1 FC 0B D9 47 45 DE D8 D8 EF 0F A0 83 9C 71 9F
35 0E 5E F1 08 63 C6 98 CC 9B 94 29 B1 A4 1A 74
5B CD 21 A8 8B F5 25 74 35 41 61 CD 97 1D 6D 4A
56 7C 44 7D 13 50 11 93 F1 AD 11 50 51 C9 37 8B
8C D8 CE 01 CB 1B BC 9E 39 F5 88 4F 9B 24 70 35
D1 AF 81 91 B1 45 44 03 07 E4 77 7B 99 3C 29 0A
81 AC AB D6 83 26 09 85 BD CD 13 CA 7E E3 D9 DA
```

tomi-font commented:
@keith-packard Would you like to get this in? Will this work with both the current SDK and the upcoming one?
@tejlmand @wearyzen Are you fine with the current changes?

keith-packard commented:
> @keith-packard Would you like to get this in? 

Yes, I'm waiting for this to finish up the SDK 1.0 integration work in Zephyr.

> Will this work with both the current SDK and the upcoming one?

Yes, I've been doing regular A/B testing using SDK 0.17.5 and a version of SDK 1.0 with a couple of additional changes to make it work with current Zephyr (those are integrated into the SDK repo, but aren't in the current SDK 1.0 beta1 version).

keith-packard commented:
> Yes, I've been doing regular A/B testing using SDK 0.17.5 and a version of SDK 1.0 with a couple of additional changes to make it work with current Zephyr (those are integrated into the SDK repo, but aren't in the current SDK 1.0 beta1 version).

With this change rebased atop the current Zephyr version (c2f9edc77f72838e7d6f5f9c0b95e4318ddfced1), I just ran `twister -a arm` using SDK 1.0 and had no failures in tests and samples using tf-m. I'll run the same using SDK 0.17.5 to verify that also works.

tomi-font commented:
> With this change rebased atop the current Zephyr version (https://github.com/zephyrproject-rtos/trusted-firmware-m/commit/c2f9edc77f72838e7d6f5f9c0b95e4318ddfced1), I just ran twister -a arm using SDK 1.0 and had no failures in tests and samples using tf-m. I'll run the same using SDK 0.17.5 to verify that also works.

Sounds good, let me know when your testing with SDK 0.17.5 is complete and let's just merge this.

keith-packard commented:
> Sounds good, let me know when your testing with SDK 0.17.5 is complete and let's just merge this.

All finished, no issues with 0.17.5 either.

tomi-font commented:
The default branch of this repository is being changed to `zephyr_tf-m_v2.2.2` which will replace `main` shortly.
Please rebase your branch atop `zephyr_tf-m_v2.2.2` and force-push it.
I have already changed your PR's base branch to `zephyr_tf-m_v2.2.2`.
Also, make sure to follow the new commit guidelines documented [here](https://github.com/zephyrproject-rtos/mbedtls?tab=readme-ov-file#additional-patches).

keith-packard commented:
> The default branch of this repository is being changed to `zephyr_tf-m_v2.2.2` which will replace `main` shortly. Please rebase your branch atop `zephyr_tf-m_v2.2.2` and force-push it. I have already changed your PR's base branch to `zephyr_tf-m_v2.2.2`. Also, make sure to follow the new commit guidelines documented [here](https://github.com/zephyrproject-rtos/mbedtls?tab=readme-ov-file#additional-patches).

Thanks. All rebased without any conflicts. I've started another Zephyr twister run "just to be sure", but I don't expect any new issues to arise.

keith-packard commented:
> Thanks. All rebased without any conflicts. I've started another Zephyr twister run "just to be sure", but I don't expect any new issues to arise.

And still passes a twister run in Zephyr. I've updated my Zephyr PR related to this (https://github.com/zephyrproject-rtos/zephyr/pull/94367)

tomi-font reviewed (approved):
Thank you, I think we can now merge this PR as it has been waiting long enough.
Be aware that my [TF-M update PR](https://github.com/zephyrproject-rtos/zephyr/pull/99248) is not yet merged in the Zephyr repository but it may go in any time. Just rebase your PR on top of it once it's in.
